models/biocro/R/run.biocro.R

```diff
@@ -24,22 +24,15 @@ run.biocro <- function(lat, lon, metpath, soil.nc = NULL, config = config, coppi
                                               10)
     config$pft$soilControl$soilDepth <- soil$ref_depth
   }
-  
-  hourly.results = list()
+
+  hourly.results <- list()
   for (i in seq_along(years)) {
     yeari <- years[i]
-    starti <- max(start.date, lubridate::ymd(paste0(yeari, "-01-01")))
-    endi <- min(end.date, lubridate::ymd(paste0(yeari, "-12-31")))
     metfile <- paste(metpath, yeari, "csv", sep = ".")
     WetDat <- data.table::fread(metfile)
-    WetDat <- WetDat[WetDat$doy >= lubridate::yday(starti) & WetDat$doy <= lubridate::yday(endi), ]
-
-    # Check that all variables are present in the expected order --
-    # BioGro accesses weather vars by position and DOES NOT check headers.
-    stopifnot(identical(colnames(WetDat), c("year", "doy", "hour", "SolarR", "Temp", "RH", "WS", "precip")))
     stopifnot(all(sapply(WetDat, is.numeric)))
-    WetDat <- as.matrix(WetDat)
-    
+    starti <- max(start.date, lubridate::ymd(paste0(yeari, "-01-01")))
+    endi <- min(end.date, lubridate::ymd(paste0(yeari, "-12-31")))
     if (!is.null(config$simulationPeriod)) {
```

Comment:
This is not a good place to sort out parameters. It requires that this function have explicit knowledge about parts of the program it shouldn't need to know about. The desired parameters should be set before calling this function.

Reply:
@justinmcgrath do you mean line 39 `stopifnot(identical(colnames(WetDat), c("year", "doy", "hour", "SolarR", "Temp", "RH", "WS", "precip")))`?

I believe that this is just an assertion to ensure that the met file meets model requirements. BioCro v 0.9x requires that a met file has this specific sequence of column names. So this could be wrapped in a conditional that only applies to BioCro v 0,9x. 

If BioCro v1.x is more flexible, there could be a different condition.

Reply:
I'm afraid I don't follow either of your comments here. Can you elaborate?

* @justinmcgrath, all the parameter checks in this section are of things that might change every year, so shouldn't the code that manages the time loop be responsible for ensuring they have sane values at each step? To what function would you suggest moving them?

* @dlebauer, you're correct that the `stopifnot(...)` is an assertion to catch misformatted met files, but it *already* applies only to BioCro 0.9x -- note that in this diff it's deleted from line 39 and added at line 103, inside the BioCro 0.9 section.

Reply:
I put this comment is a stupid spot. This refers to the if else statements that select day1 and dayn.

The function needs to know day1 and dayn, but it doesn't need to sort out where it's getting them from. An easier to use design is for the function to accept only one set of parameters, and the caller should decide what parameters to use. You could set day1 and dayn from `config`, get rid of the other choices, and make it an error to not provide day1 and dayn. The same reasoning applies to the use of soil.nc at the beginning.

The order of columns does not matter in the new version. Variables are always identified by their names. Not supplying a required name is an error, and it should list the missing names. There aren't built in functions in C++ for that, so sometimes it just gives a generic error if I missed something.

Reply:
@infotroph my comment is that for BioCro v0.9, the column names do not matter, only the position does. Therefore, if there is a single BioCro met standard that uses v1.0 names and 0.9 order then this condition could just test that the dataset has v1.0 names in v0.9 order. (PEcAn never needs to know about the names WS and  SolarR that only appear in a few sample datasets, and should not be stopped if it has a dataset with v1 names).

Reply:
Column names fixed in 4604149501c4cf17b31607ea919a42fdc3b5eccc